src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let previousState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        previousState = createStateWithAddedProperties(
+          previousState,
+          action.extraData,
+        );
+        break;
+      case 'removeProperties':
+        previousState = createStateWithRemovedProperties(
+          previousState,
+          action.keysToRemove,
+        );
+        break;
+      case 'clear':
+        previousState = createClearState(previousState);
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    history.push({ ...previousState });
```

Comment:
You are pushing a new object copy into the history array after each action. However, this operation is common for all cases and should be performed once after the switch statement to adhere to the DRY principle.